tools/DeleteModules.ps1

```diff
@@ -0,0 +1,208 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ----------------------------------------------------------------------------------
+
+#Requires -Modules AzureRM.KeyVault
+[CmdletBinding(
+    DefaultParameterSetName='Scope', 
+    SupportsShouldProcess=$true
+)]
+param(
+    [Parameter(ParameterSetName='Scope', Mandatory = $true, Position = 0)]
+    [ValidateSet("AzureRMAndDependencies", "AzureAndDependencies", "NetCoreModules", "AzureStackAndDependencies")]
+    [string] $scope,
+    [Parameter(ParameterSetName='ModuleList', Mandatory = $false)]
+    [string[]] $listOfModules,
+    [Parameter(Mandatory = $false)]
+    [ValidateSet("TestGallery", "PSGallery")]
+    [string] $repoName,
+    [Parameter(Mandatory = $false)]
+    [string] $nugetExe,
+    [Parameter(Mandatory = $false)]
+    [string] $apiKey
+)
+
+function Get-TargetModules
+{
+    [CmdletBinding()]
+    param
+    (
+      [string]$scope,
+      [string[]]$moduleList
+    )
+
+    if ($listOfModules.Count -ge 1) 
+    {
+        $targets = @()
+        $moduleList | ForEach-Object {
+            $targets += Find-Module -Name $_ -Repository $repoName
+        }
+
+        return $targets
+    }
+
+    else
+    {
+        $targets = @()
+        $query = ""
+        if ($scope -eq "AzureRMAndDependencies") {
+            $query = "AzureRM"
+        } elseif ($scope -eq "AzureAndDependencies") {
+            $query = "Azure"
+        } elseif ($scope -eq "NetCoreModules") {
+            $query = "AzureRM.Netcore"
+        } elseif ($scope -eq "AzureStackAndDependencies") {
+            $query = "AzureStack"
+        }
+
+        $azureRmAllVersions = Find-Module -Name $query -Repository PSGallery -AllVersions
+        $targets += $azureRmAllVersions[0]
+        $currentDependencies = $azureRmAllVersions[0].Dependencies
+        $previousDependencies = $azureRmAllVersions[1].Dependencies
+        $currentDependencies | ForEach-Object {
+            $CDModule = $_
+            $versionChanged = $true
+            $previousDependencies | ForEach-Object {
+                if (($_.Name -eq $CDModule.Name) -and (($_.RequiredVersion -eq $CDModule.RequiredVersion) -and ($_.MinimumVersion -eq $CDModule.MinimumVersion)))
+                {
+                    $versionChanged = $false
+                }
+            }
+
+            if ($versionChanged)
+            {
+                $targets += $CDModule
+            }
+        }
+
+        return $targets
+    }
+}
+
+function Get-DependentModules
+{
+    param(
+        [string]$repoName,
+        [string]$moduleName,
+        [string]$moduleVersion,
+        $allModules,
+        $azureModules
+    )
+
+    $dependencies = @()
+    $azureModules | ForEach-Object {
+        $targetName = $_.Name
+        $moduleNotBeingDeleted = $true
+        $allModules | ForEach-Object {
+            if ($_.Name -eq $targetName)
+            {
+                $moduleNotBeingDeleted = $false
```

Comment:
Also, I think this would be easier to follow if we used ```$allModules | Where {$_.Name -eq $targetName}``` to get the set of elements matching target name

Reply:
Changed